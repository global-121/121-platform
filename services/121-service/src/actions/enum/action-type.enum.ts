export enum ActionType {
  startBlockNewPayment = 'start-block-new-payment',
  endBlockNewPayment = 'end-block-new-payment',
}
