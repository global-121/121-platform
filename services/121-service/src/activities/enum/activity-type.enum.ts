export enum ActivityTypeEnum {
  DataChange = 'data-change',
  FinancialServiceProviderChange = 'financial-service-provider-change',
  Message = 'message',
  Note = 'note',
  StatusChange = 'status-change',
  Transaction = 'transaction',
}
