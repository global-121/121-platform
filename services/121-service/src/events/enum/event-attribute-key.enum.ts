export enum EventAttributeKeyEnum {
  reason = 'reason',
  oldValue = 'oldValue',
  newValue = 'newValue',
  fieldName = 'fieldName',
  duplicateWithRegistrationId = 'duplicateWithRegistrationId',
  duplicateWithRegistrationProgramId = 'duplicateWithRegistrationProgramId',
}
