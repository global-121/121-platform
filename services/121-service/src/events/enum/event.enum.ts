export enum EventEnum {
  registrationDataChange = 'registrationDataChange',
  financialServiceProviderChange = 'financialServiceProviderChange',
  registrationStatusChange = 'registrationStatusChange',
  ignoredDuplicate = 'ignoredDuplicate',
}
