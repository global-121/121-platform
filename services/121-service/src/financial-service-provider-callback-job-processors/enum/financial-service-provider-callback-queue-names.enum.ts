export enum FinancialServiceProviderCallbackQueueNames {
  safaricomTransferCallback = 'safaricomTransferCallback',
  safaricomTimeoutCallback = 'safaricomTimeoutCallback',
}
