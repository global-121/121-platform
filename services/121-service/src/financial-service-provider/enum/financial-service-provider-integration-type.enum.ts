export enum FinancialServiceProviderIntegrationType {
  api = 'api',
  csv = 'csv',
  xml = 'xml',
}
