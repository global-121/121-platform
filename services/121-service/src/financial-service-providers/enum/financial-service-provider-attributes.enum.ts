export enum FinancialServiceProviderAttributes {
  phoneNumber = 'phoneNumber',
  nationalId = 'nationalId',
  fullName = 'fullName',
  addressStreet = 'addressStreet',
  addressHouseNumber = 'addressHouseNumber',
  addressHouseNumberAddition = 'addressHouseNumberAddition',
  addressPostalCode = 'addressPostalCode',
  addressCity = 'addressCity',
  bankAccountNumber = 'bankAccountNumber',
  whatsappPhoneNumber = 'whatsappPhoneNumber',
}
