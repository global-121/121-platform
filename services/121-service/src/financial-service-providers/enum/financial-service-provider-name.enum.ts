export enum FinancialServiceProviderName {
  intersolveVoucherWhatsapp = 'Intersolve-voucher-whatsapp',
  intersolveVoucherPaper = 'Intersolve-voucher-paper',
  intersolveVisa = 'Intersolve-visa',
  intersolveJumboPhysical = 'Intersolve-jumbo-physical',
  africasTalking = 'Africas-talking',
  belcash = 'BelCash',
  vodacash = 'VodaCash',
  bobFinance = 'BoB-finance',
  ukrPoshta = 'UkrPoshta',
  safaricom = 'Safaricom',
  commercialBankEthiopia = 'Commercial-bank-ethiopia',
  excel = 'Excel',
}

export enum FinancialServiceProviderConfigurationEnum {
  password = 'password',
  username = 'username',
  columnsToExport = 'columnsToExport',
  columnToMatch = 'columnToMatch',
  brandCode = 'brandCode',
  displayName = 'displayName',
  coverLetterCode = 'coverLetterCode',
}

export const FinancialServiceProviderConfigurationMapping: {
  [key in FinancialServiceProviderName]?: any;
} = {
  [FinancialServiceProviderName.intersolveVoucherWhatsapp]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveVoucherPaper]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveVisa]: [
    FinancialServiceProviderConfigurationEnum.brandCode,
    FinancialServiceProviderConfigurationEnum.coverLetterCode,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.intersolveJumboPhysical]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.africasTalking]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.belcash]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.vodacash]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.bobFinance]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.ukrPoshta]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.safaricom]: [
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.commercialBankEthiopia]: [
    FinancialServiceProviderConfigurationEnum.password,
    FinancialServiceProviderConfigurationEnum.username,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
  [FinancialServiceProviderName.excel]: [
    FinancialServiceProviderConfigurationEnum.columnsToExport,
    FinancialServiceProviderConfigurationEnum.columnToMatch,
    FinancialServiceProviderConfigurationEnum.displayName,
  ],
};
