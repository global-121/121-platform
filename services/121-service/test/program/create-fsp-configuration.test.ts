/* eslint-disable jest/no-conditional-expect */
import { HttpStatus } from '@nestjs/common';

import { SeedScript } from '@121-service/src/scripts/seed-script.enum';
import programOCW from '@121-service/src/seed-data/program/program-nlrc-ocw.json';
import {
  getProgram,
  postProgram,
} from '@121-service/test/helpers/program.helper';
import {
  cleanProgramForAssertions,
  getAccessToken,
  resetDB,
} from '@121-service/test/helpers/utility.helper';

describe('Create program questions', () => {
  let accessToken: string;

  beforeEach(async () => {
    await resetDB(SeedScript.nlrcMultiple);
    accessToken = await getAccessToken();
  });

  it('should create a program with FSP configuration', async () => {
    // Arrange
    const program = JSON.parse(JSON.stringify(programOCW));

    // Add test display name to program
    const intersolveVoucherWhatsappTranslations = {
      name: 'displayName',
      value: {
        en: 'Intersolve Voucher WhatsApp',
        nl: 'Intersolve Voucher WhatsApp Dutch translation',
        es: 'Intersolve Voucher WhatsApp Spanish translation',
      },
    };

    program.financialServiceProviders[0].configuration.push(
      intersolveVoucherWhatsappTranslations,
    );

    // Act
    const createProgramResponse = await postProgram(program, accessToken);

    // Assert
    const programId = createProgramResponse.body.id;
    const getProgramResponse = await getProgram(programId, accessToken);
    expect(createProgramResponse.statusCode).toBe(HttpStatus.CREATED);

    const cleanedProgram = cleanProgramForAssertions(program);
    const cleanedProgramResponse = cleanProgramForAssertions(
      getProgramResponse.body,
    );

    expect(cleanedProgramResponse).toMatchSnapshot(
      `Create program response for program: ${program.titlePortal.en}`,
    );

    expect(cleanedProgramResponse).toMatchObject(cleanedProgram);
  });
});
