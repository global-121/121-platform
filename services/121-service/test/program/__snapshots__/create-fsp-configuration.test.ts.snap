// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create program questions should create a program with FSP configuration: Create program response for program: NLRC OCW program 1`] = `
{
  "aboutProgram": {
    "en": "This is the OCW program.",
  },
  "allowEmptyPhoneNumber": false,
  "budget": null,
  "currency": "EUR",
  "deprecatedCustomDataKeys": [],
  "description": {
    "en": "",
  },
  "distributionDuration": 46,
  "distributionFrequency": "2-weeks",
  "editableAttributes": [
    {
      "label": {
        "en": "First name",
      },
      "name": "firstName",
      "type": "text",
    },
    {
      "label": {
        "en": "Last name",
      },
      "name": "lastName",
      "type": "text",
    },
  ],
  "enableMaxPayments": false,
  "enableScope": false,
  "filterableAttributes": [
    {
      "filters": [
        {
          "allowedOperators": [
            "$eq",
            "$in",
            "$ilike",
            "$null",
          ],
          "isInteger": false,
          "name": "lastMessageStatus",
        },
      ],
      "group": "messages",
    },
    {
      "filters": [
        {
          "allowedOperators": [
            "$eq",
            "$in",
            "$ilike",
            "$null",
          ],
          "isInteger": false,
          "name": "addressCity",
        },
        {
          "allowedOperators": [
            "$eq",
            "$in",
            "$ilike",
            "$null",
          ],
          "isInteger": false,
          "name": "addressHouseNumber",
        },
        {
          "allowedOperators": [
            "$eq",
            "$in",
            "$ilike",
            "$null",
          ],
          "isInteger": false,
          "name": "addressHouseNumberAddition",
        },
        {
          "allowedOperators": [
            "$eq",
            "$in",
            "$ilike",
            "$null",
          ],
          "isInteger": false,
          "name": "addressPostalCode",
        },
        {
          "allowedOperators": [
            "$eq",
            "$in",
            "$ilike",
            "$null",
          ],
          "isInteger": false,
          "name": "addressStreet",
        },
        {
          "allowedOperators": [
            "$eq",
            "$in",
            "$ilike",
            "$null",
          ],
          "isInteger": false,
          "name": "financialServiceProvider",
        },
        {
          "allowedOperators": [
            "$eq",
            "$in",
            "$ilike",
            "$null",
          ],
          "isInteger": false,
          "name": "firstName",
        },
        {
          "allowedOperators": [
            "$eq",
            "$null",
          ],
          "isInteger": true,
          "name": "inclusionScore",
        },
        {
          "allowedOperators": [
            "$eq",
            "$in",
            "$ilike",
            "$null",
          ],
          "isInteger": false,
          "name": "lastName",
        },
        {
          "allowedOperators": [
            "$eq",
            "$null",
          ],
          "isInteger": true,
          "name": "paymentAmountMultiplier",
        },
        {
          "allowedOperators": [
            "$eq",
            "$in",
            "$ilike",
            "$null",
          ],
          "isInteger": false,
          "name": "personAffectedSequence",
        },
        {
          "allowedOperators": [
            "$eq",
            "$in",
            "$ilike",
            "$null",
          ],
          "isInteger": false,
          "name": "phoneNumber",
        },
        {
          "allowedOperators": [
            "$eq",
            "$in",
            "$ilike",
            "$null",
          ],
          "isInteger": false,
          "name": "preferredLanguage",
        },
        {
          "allowedOperators": [
            "$eq",
            "$in",
            "$ilike",
            "$null",
          ],
          "isInteger": false,
          "name": "referenceId",
        },
        {
          "allowedOperators": [
            "$eq",
            "$in",
            "$ilike",
            "$null",
          ],
          "isInteger": false,
          "name": "registrationCreatedDate",
        },
        {
          "allowedOperators": [
            "$eq",
            "$in",
            "$ilike",
            "$null",
          ],
          "isInteger": false,
          "name": "whatsappPhoneNumber",
        },
      ],
      "group": "paAttributes",
    },
    {
      "filters": [
        {
          "allowedOperators": [
            "$eq",
            "$null",
          ],
          "isInteger": true,
          "name": "failedPayment",
        },
        {
          "allowedOperators": [
            "$eq",
            "$null",
          ],
          "isInteger": true,
          "name": "notYetSentPayment",
        },
        {
          "allowedOperators": [
            "$eq",
            "$null",
          ],
          "isInteger": true,
          "name": "paymentCount",
        },
        {
          "allowedOperators": [
            "$eq",
            "$null",
          ],
          "isInteger": true,
          "name": "successPayment",
        },
        {
          "allowedOperators": [
            "$eq",
            "$null",
          ],
          "isInteger": true,
          "name": "waitingPayment",
        },
      ],
      "group": "payments",
    },
  ],
  "financialServiceProviders": [
    {
      "displayName": {
        "en": "Visa debit card",
      },
      "fsp": "Intersolve-visa",
      "hasReconciliation": false,
      "id": 3,
      "integrationType": "api",
      "notifyOnTransaction": true,
      "questions": [
        {
          "answerType": "text",
          "duplicateCheck": false,
          "export": [
            "all-people-affected",
            "included",
          ],
          "fspId": 3,
          "id": 6,
          "label": {
            "en": "Address city",
          },
          "name": "addressCity",
          "options": null,
          "pattern": ".+",
          "placeholder": null,
          "showInPeopleAffectedTable": false,
        },
        {
          "answerType": "numeric",
          "duplicateCheck": false,
          "export": [
            "all-people-affected",
            "included",
          ],
          "fspId": 3,
          "id": 3,
          "label": {
            "en": "Address house number",
          },
          "name": "addressHouseNumber",
          "options": null,
          "pattern": null,
          "placeholder": null,
          "showInPeopleAffectedTable": false,
        },
        {
          "answerType": "text",
          "duplicateCheck": false,
          "export": [
            "all-people-affected",
            "included",
          ],
          "fspId": 3,
          "id": 4,
          "label": {
            "en": "Address house number addition",
          },
          "name": "addressHouseNumberAddition",
          "options": null,
          "pattern": null,
          "placeholder": null,
          "showInPeopleAffectedTable": false,
        },
        {
          "answerType": "text",
          "duplicateCheck": false,
          "export": [
            "all-people-affected",
            "included",
          ],
          "fspId": 3,
          "id": 5,
          "label": {
            "en": "Address postal code",
          },
          "name": "addressPostalCode",
          "options": null,
          "pattern": ".+",
          "placeholder": null,
          "showInPeopleAffectedTable": false,
        },
        {
          "answerType": "text",
          "duplicateCheck": false,
          "export": [
            "all-people-affected",
            "included",
          ],
          "fspId": 3,
          "id": 2,
          "label": {
            "en": "Address street",
          },
          "name": "addressStreet",
          "options": null,
          "pattern": ".+",
          "placeholder": null,
          "showInPeopleAffectedTable": false,
        },
        {
          "answerType": "tel",
          "duplicateCheck": true,
          "export": [
            "all-people-affected",
            "included",
          ],
          "fspId": 3,
          "id": 7,
          "label": {
            "en": "WhatsApp Nr.",
          },
          "name": "whatsappPhoneNumber",
          "options": null,
          "pattern": null,
          "placeholder": {
            "ar": "00 00 00 00 0 00+",
            "en": "+00 0 00 00 00 00",
          },
          "showInPeopleAffectedTable": false,
        },
      ],
    },
    {
      "displayName": {
        "en": "Intersolve Voucher WhatsApp",
        "es": "Intersolve Voucher WhatsApp Spanish translation",
        "nl": "Intersolve Voucher WhatsApp Dutch translation",
      },
      "fsp": "Intersolve-voucher-whatsapp",
      "hasReconciliation": false,
      "id": 1,
      "integrationType": "api",
      "notifyOnTransaction": false,
      "questions": [
        {
          "answerType": "tel",
          "duplicateCheck": true,
          "export": [
            "all-people-affected",
            "included",
          ],
          "fspId": 1,
          "id": 1,
          "label": {
            "en": "WhatsApp Nr.",
          },
          "name": "whatsappPhoneNumber",
          "options": null,
          "pattern": null,
          "placeholder": {
            "ar": "00 00 00 00 0 00+",
            "en": "+00 0 00 00 00 00",
          },
          "showInPeopleAffectedTable": true,
        },
      ],
    },
  ],
  "fixedTransferValue": 25,
  "fullnameNamingConvention": [
    "firstName",
    "lastName",
  ],
  "id": 4,
  "languages": [
    "en",
    "nl",
    "ar",
    "tr",
  ],
  "location": "Nederland",
  "monitoringDashboardUrl": "https://app.powerbi.com/view?r=eyJrIjoiNDkxN2Q1NmEtOWIxOC00ZTYyLTkxMTMtYmI3MmE4YTVkMTE2IiwidCI6ImEyYjUzYmU1LTczNGUtNGU2Yy1hYjBkLWQxODRmNjBmZDkxNyIsImMiOjh9",
  "ngo": "NLRC",
  "paTableAttributes": [
    {
      "label": {
        "en": "Address city",
      },
      "name": "addressCity",
      "questionType": "fspQuestion",
      "type": "text",
    },
    {
      "label": {
        "en": "Address house number",
      },
      "name": "addressHouseNumber",
      "questionType": "fspQuestion",
      "type": "numeric",
    },
    {
      "label": {
        "en": "Address house number addition",
      },
      "name": "addressHouseNumberAddition",
      "questionType": "fspQuestion",
      "type": "text",
    },
    {
      "label": {
        "en": "Address postal code",
      },
      "name": "addressPostalCode",
      "questionType": "fspQuestion",
      "type": "text",
    },
    {
      "label": {
        "en": "Address street",
      },
      "name": "addressStreet",
      "questionType": "fspQuestion",
      "type": "text",
    },
    {
      "label": {
        "en": "First name",
      },
      "name": "firstName",
      "questionType": "programCustomAttribute",
      "type": "text",
    },
    {
      "label": {
        "en": "Last name",
      },
      "name": "lastName",
      "questionType": "programCustomAttribute",
      "type": "text",
    },
    {
      "label": {
        "en": "Phone Number",
      },
      "name": "phoneNumber",
      "questionType": "programQuestion",
      "type": "tel",
    },
    {
      "label": {
        "en": "WhatsApp Nr.",
      },
      "name": "whatsappPhoneNumber",
      "questionType": "fspQuestion",
      "type": "tel",
    },
    {
      "label": {
        "en": "WhatsApp Nr.",
      },
      "name": "whatsappPhoneNumber",
      "questionType": "fspQuestion",
      "type": "tel",
    },
  ],
  "paymentAmountMultiplierFormula": null,
  "programCustomAttributes": [
    {
      "duplicateCheck": false,
      "id": 4,
      "label": {
        "en": "First name",
      },
      "name": "firstName",
      "programId": 4,
      "showInPeopleAffectedTable": false,
      "type": "text",
    },
    {
      "duplicateCheck": false,
      "id": 5,
      "label": {
        "en": "Last name",
      },
      "name": "lastName",
      "programId": 4,
      "showInPeopleAffectedTable": false,
      "type": "text",
    },
  ],
  "programQuestions": [
    {
      "answerType": "tel",
      "duplicateCheck": true,
      "editableInPortal": false,
      "export": [],
      "id": 4,
      "label": {
        "en": "Phone Number",
      },
      "name": "phoneNumber",
      "options": null,
      "pattern": null,
      "persistence": true,
      "placeholder": {
        "en": "+31 6 00 00 00 00",
      },
      "programId": 4,
      "questionType": "standard",
      "scoring": {},
      "showInPeopleAffectedTable": false,
    },
  ],
  "published": true,
  "targetNrRegistrations": 100000,
  "titlePortal": {
    "en": "NLRC OCW program",
  },
  "tryWhatsAppFirst": true,
  "validation": false,
}
`;
